### ISSUE TITLE
Getting inverse units

### ISSUE THREAD

**Issue by ypearson** [bernedom/SI#101]
#include <SI/length.h>

using namespace SI::literals;
constexpr auto pitch = 2 / 1_mm;

**Expected behavior**
I expected to pitch to be of type 2 mm exp -1 (2 per millimeter) ratio of -1

I/include/SI/detail/unit.h:474:33: error: ‘(2 / 0)’ is not a constant expression

**Comment by ypearson** [comment]
Possible workaround:
`using threadPitch_unit_mm = struct SI::detail::unit_t<'L', std::ratio_multiply<std::ratio<-1L>, std::ratio<1L>>, int64_t, std::milli>;
constexpr threadPitch_unit_mm threadPitch_per_mm{2};
`

**Comment by ypearson** [comment]
`constexpr auto x = 2.0 / SI::milli_metre_t<double>{1}; //Works`
`constexpr auto x = 2 / SI::milli_metre_t<double>{1}; // Doesn't work`

**Comment by bernedom** [comment]
It seems the corresponding operator is missing from unit.h

Will add it.

**Comment by ypearson** [comment]
Thank you! Appreciate it very much

**Comment by bernedom** [comment]
On a closer look, you might have found a genuine bug/weird behavior of the implementation here, as the resulting value is . To clarify, assuming the following calculation
```
int64_t v{1000}
const auto mm = 2_mm;
const auto r = v / mm;
```
The expected result would be of type `mm^-1` and the value of it would be 500.

**Comment by ypearson** [comment]
I'm glad you're seeing what I'm seeing. For my application I need 1/mm for
describing a mechanical interface.

Do you think the workaround I posted is okay for right now?

Thanks for your attention to this.

**Comment by bernedom** [comment]
Yes the workaround might work. But there might be a bug in the operator at the momeent, so use it with care and check the behavior first.

**Comment by ypearson** [comment]
Okay thank you. This is highly appreciated, I'm using your library for some
really important work. Keep me posted!:)

**Comment by bernedom** [comment]
The issue is resolved in this branch https://github.com/bernedom/SI/tree/bugfix/fix-compilation-failure-for-constexpr-operator and will be in release `2.4.1` 

Could you please check it quickly, if it works as you expect and give me feedback? 
```cpp
constexpr auto pitch = 2 / 1_mm; // works 
constexpr auto x = 2.0 / SI::milli_metre_t<double>{1};  // works 
constexpr auto x = 2 / SI::milli_metre_t<double>{1}; // works but will produce a compiler warning because of the narrowing from int to double
```

**Comment by ypearson** [comment]
Absolutely, I'm traveling at the moment. But by next week I'll be at my
desk looking at this carefully.

**Comment by ypearson** [comment]
Back online, looking at the fix today

**Comment by bernedom** [comment]
The C++ standard only offers `unsigned long long int` and `long double` and various `char_t` as arguments for user defined literals for. So for numerical literals it's either `long long int` or `long double` as type. 
Since SI should support signed values choosing `int64_t` (or `long long int` on 64bit machines) as default storage type for integrals was a kind of natural choice. I thought about creating literals like `_mmF` which would return a `float` but then the number of literals would explode without too much benefit. There currently exists no way to templatize 

In the current implementation of `int / unit<int64_t>` will return a  `unit<int>`, but I'm planning to implement the standard arithmetic conversions as soon as I find the time as described here: https://en.cppreference.com/w/c/language/conversion (This will not be part of the current fix for the original problem)

In the end I try to keep as close as possible to the behavior as when calculating with raw numbers. 
More details about narrowing and possible loss of precision is described here: 
https://github.com/bernedom/SI/blob/main/doc/implementation-details.md#implicit-ratio-conversion--possible-loss-of-precision

As for the second calculation `1_mm/2` will do an integer division so 0 is what is expected. Automatically switching types would work against the strong type safety intended and also increase the calculation overhead way too much, because each time the result has to be checked and possibly calculated a second time.

**Comment by ypearson** [comment]
Thanks for the response. Makes sense to me!
I just realized I can do `1.0_mm/2` to get the result I was looking for
